PluginExampleProject/src/main/scala/com/alpine/plugin/samples/ver1_0/KMeansModel.scala

```diff
@@ -0,0 +1,234 @@
+    })
+
+    val k = operatorParameters.getIntValue(KMeansConstants.numClustersParamId)
+    val n = operatorParameters.getIntValue(KMeansConstants.numIterationsParamId)
+
+    val mllibModel: KMeansModel = KMeans.train(vectorRDD, k, n )
+
+    listener.notifyMessage("The model was trained with " + k + " clusters and "  + n + " iterations")
+
+    val clusters : Array[DenseVector] = mllibModel.clusterCenters.map(_.toDense)
+
+    //filter the input column definitions base on the value of the input features parameters
+    //in order to get a sequence of the full Alpine Column Definitions  for the input feature
+    val inputFeatureColumnDefs: Array[ColumnDef] =
+      input.tabularSchema.getDefinedColumns.filter( columnDef =>
+     inputFeatureNames.contains(columnDef.columnName)).toArray
+
+    val clusterWithName = clusters.zipWithIndex.map{
+      case (vector, index) => ("Cluster_"+index)
+    }
+
+    val labels = clusterWithName.toSeq
+    val clustersAsDoubleArrays = clusters.map(_.toArray)
+
+
+    val alpineModel = new ExampleKMeansClusteringModel(inputFeatureColumnDefs,
+      clustersAsDoubleArrays, labels)
+
+    //this will be used in the visualization to create a tab that shows which clusters were trained
+    val clusterAddendum = makeAddendum(inputFeatureNames, clusters, labels)
+
+    new ClusteringModelWrapper(modelName = "K Means Model", model = alpineModel, sourceOperatorInfo =
+    Some(operatorParameters.operatorInfo), addendum = clusterAddendum)
+  }
+
+  /**
+   * Create the Map used in the "addendum" in the creation of the Clustering Model Wrapper.
+   * The information will be avalible at output visualization, so we can use it to display
+   * a table with information about the centers.
+   */
+
+  def makeAddendum(inputFeatureNames : Array[String], clusters : Array[DenseVector],
+    labels : Seq[String]) = {
+
+    val clusterTableHeader = "ClusterName" :: inputFeatureNames.toList
+
+    val clusterTableBody = labels.zip(clusters).map{
+      case (name, vector) => name ::  vector.values.map(_.toString).toList
+    }
+    val clusterTable = HtmlTabulator.format(clusterTableHeader :: clusterTableBody.toList )
+
+    Map[String, AnyRef](KMeansConstants.visualOutputKey -> clusterTable)
+  }
+}
+
+/**
+ * Defines the Clustering model, extends the Alpine ClusteringRowModel class.
+ * Note: I have designed this class so that it takes some parameters
+ * (which we get from the parameters of the Operator and from the from the MLLib Model)
+ * but the uer may design the class so that it takes any parameter that are convenient).
+ *
+ * NOTE: We have an existing K Means Model and transformer included in the Model SDK. They
+ * are called KMeansModel and KMeansModelTransformer. However this sample is intended to show how
+ * one could write more custom code in a transformation model.
+ *
+ * NOTE: In order to use this with the existing Alpine Predictor, this class CANNOT contain any Spark Dependencies.
```

Comment:
It might be healthier to have these in separate .scala files, so we can see from the imports that there are no Spark dependencies.